### DISCUSSION TITLE
DICOM SPACING PROBLEM

### DISCUSSION THREAD

**Discussion by oscarramirezs** [Project-MONAI/MONAI#1637] [General]
Hi,

Thank you for the nice library! Sorry I'm new to MONAI, but I'm stock, can anyone help me? please

I am dealing with CT's (DICOM fortmat) => (Deep, 512, 512) that means (Z, X, Y) deep is the Z axis and it Deep change switches between each CT for example (172, 512, 512) and the spacing ( X, Y, Z) for example [0.3984375 0.3984375 1. ]

So, using the  AddChanneld (keys = ["image", "label"])

=> This is the output: image shape(1, 172, 512, 512) which is fine

Then using Spacingd (keys = ["image", "label"], pixdim = (
              2.0, 1.5, 1.5), mode = ("bilinear", "closest")),

=> This is the output: image shape (172, 103, 137) which is bad

What can I do to fix this and get X and Y the same size?

THANKS AGAIN,

I don't know if I am clear enough

Oscar

**Comment by rijobro** [comment]
I'm confused by your dimension ordering. In your image, you mention `(z, y, x)`, but for your spacing you gave `(x, y, z)`. Would doing `spacing[::-1]` solve your problem? This would reverse the order of your spacing to also be `(z, y, x)`.

**Reply by oscarramirezs** [reply]
Hi thank you for your help 

printing => image, metadata = loader(train_files[0]["image"])

input: Images/5D131CC56CAA4_A_mimics.mhd
image shape: (172, 512, 512)
image affine:
[[ 3.98437500e-01  0.00000000e+00  0.00000000e+00  4.41992188e+01]
 [ 0.00000000e+00  3.98437500e-01  0.00000000e+00 -2.20800781e+02]
 [ 0.00000000e+00  0.00000000e+00  1.00000000e+00  7.32799988e+02]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]]
spacing:
[0.3984375 0.3984375 1.       ]
origin:
[  44.19921875 -220.80078125  732.79998779]
direction:
[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


and after spacing = Spacingd(keys=["image", "label"], pixdim=(2, 1.5, 1.5), mode=("bilinear", "nearest")) I had this 

image shape: (172, 103, 137)
label shape: (172, 103, 137)
image affine after Spacing:
[[   2.            0.            0.           44.19921875]
 [   0.            1.5           0.         -220.80078125]
 [   0.            0.            1.          732.79998779]
 [   0.            0.            0.            1.        ]]
label affine after Spacing:
[[   2.            0.            0.           44.19921875]
 [   0.            1.5           0.         -220.80078125]
 [   0.            0.            1.          732.79998779]
 [   0.            0.            0.            1.        ]]
Spacing:
[0.3984375 0.3984375 1.       ]

**Reply by wyli** [reply]
I think there might be something wrong with the original input, the spacing is `[0.3984375 0.3984375 1. ]` but image shape is `shape: (172, 512, 512)` this is quite unusual...

**Comment by oscarramirezs** [comment]
Sorry but i'm really confused

1. why in the example "Spleen 3D segmentation with MONAI"  after appling Setup transforms for training and validation,

 the output of Check transforms in DataLoader is
 image shape: torch.Size ([226, 157, 113]), label shape: torch.Size([226, 157, 113]).
 
I expected a square image with a deep, . if the original images is 512 512 deep matrix, I m wrong? 

![image](https://user-images.githubusercontent.com/52659349/109167875-896ae200-777e-11eb-9d42-3cb1daf8c95d.png)

**Reply by rijobro** [reply]
The validation transforms for the spleen example use `CropForegroundd`, which crops images as small as possible without removing any non-zero voxels. You can see this in the screenshot that you've attached above -- the image is cropped tightly around the "image" key (on the left). 

I think this is unrelated to your problem. I think your problem is that there is a mix up between your x and z dimensions. It says that the number of voxels is equal in the x and y dimensions, but the spacing is equal in the y and z dimensions.

If you open your DICOMs in a viewer (e.g., Slicer, ITKSnap), does the aspect ratio of the images look correct?

**Reply by oscarramirezs** [reply]
thank 

I tried the in local the  spleen example whitout CropForegroundd, and the output is not a square, 

about my data, it looks ok,

Is there any way that I can share with you,  two or three examples of my data, It would be a great help, please,
